### ISSUE TITLE
No echo in server

### ISSUE THREAD

**Issue by Hagb** [MuhdNurHidayat/unicodeparser#3]
I use the mod send text,and I will not be able to see what I say, while other people in server can see.
But if I send text by original Minetest way, there will be no trouble.
What's more, if I use my own server in local, there will be no trouble too.

**Comment by mnh48** [comment]
Did you set Minetest advanced setting to use CJK font such as Noto Sans CJK SC?
This is because the default font in Minetest can't display Chinese, Japanese and Korean (CJK) text, you need to change the font to another font that is compatible with CJK.

Noto Sans CJK font can be download here: https://www.google.com/get/noto/help/cjk/

If changing font didn't work, maybe put some screenshot of what happened so I could check what's wrong?

**Comment by Hagb** [comment]
@MuhdNurHidayat I had set it, and the problem is not fonts.
I can't not see anything like "\<myname\> xxxx" if I use ".uc" command to send text in the server "mc.tianshuyu.cc:30000", but other player can see.
**If I do it when connecting to my own server in local, there will be no trouble.**  So I don't know what problem is now.

**Comment by mnh48** [comment]
the ".uc" command just use the default send message API in client side mod, it appear fine on my side even on multiplayer so I'm not sure what's wrong on your side.

This is test on multiplayer server:
![2018-07-30_03-53-11](https://user-images.githubusercontent.com/12217018/43370135-3d411056-93ac-11e8-99da-84f8091f95af.gif)

Can I know which server you play on, or the server version?
And your client version too.

**Comment by Hagb** [comment]
@MuhdNurHidayat 
I find that it maybe happens only in "mc.tianshuyu.cc:" with default port 30000, in which the text sent by the CSM no matter only including English or including Chinese has the same problem.
I have also tried other servers..... Only in "mc.tianshuyu.cc" itcan happen...
It is strange that all people except me in server "mc.tianshuyu.cc" can see what I said.

**Comment by mnh48** [comment]
just like what I thought... because that happened in HOMETOWN in the old days when the server is not updated, it's now working as expected as the server is now 0.4.17.1 ...

So in your case, it looks like the server itself is running old version of minetest, they need to update it...

You can try to disable client mod and just chat normally in english on the server, you will notice english text also did not appeared, even when client mod are disabled...

---
Explanation of what actually happen:
In older minetest, the client will print the player's message before sending to server so server don't display them. This make it confuse if the message did not reach the server (eg. lagging). When player wrote it in chat and saw their message, they thought other players already receive the chat as well even when it's not received yet.

In newer minetest, to fix the earlier issue, they move the function to server... so when player send message, the client did not display anything, but the server will display it once it arrives

What happen to you is because you use the new client on old server, new client do not print on client anymore because it expect the server to display it for them... but you connect to old server where old server will not display what you write because old client will already print what was written before sending to server.

If you connect old client to new server instead then you might see double message from your side as both sides are printing the message.

**Comment by Hagb** [comment]
Isn't like what you said. It is strange that **every text sent by unicodeparser has no echo but every text not sent by unicodeparser way has echo**. （用unicodeparser发的消息是没有回显的，用原始方法发的是没有回显的，不管是不是全英文） Of course, it is only in that server.（只有那个服务器有这个现象）
Thanks.

**Comment by mnh48** [comment]
That might have something to do with the client-side mods API then...
As I said earlier, unicodeparser only use the [client-side mods API](https://github.com/minetest/minetest/blob/stable-0.4/doc/client_lua_api.md) provided by Minetest itself, specifically it use the function `minetest.send_chat_message(message)`.

Can you look for the server's minetest version?
The server version is shown when you send `/status` in chat.
If I know the server version, I might be able to do workaround on the API limitation.

Any server should show their version after you send `/status` in chat, if not then it might be a custom server which I can't help with.
![status](https://user-images.githubusercontent.com/12217018/43441952-1211ce06-94cf-11e8-913c-e3127f69048f.png)
^ in this example, the server is running version 0.4.16

**Comment by Hagb** [comment]
The version of server is 0.4.15 .
And my client is 0.4.17.1 .

**Comment by Hagb** [comment]
![2](https://user-images.githubusercontent.com/10961094/43443315-17003f52-94d3-11e8-89f6-bad7bc8cf106.gif)
Noto: I use UnicodeConverter.sh and the echo "you said: test" is from #5 .

**Comment by mnh48** [comment]
After some testing, I find out what happened to you actually happened on all servers with version 0.4.15 and below (server protocol version 28 and below). 

I just merged 1baee5e which checks for server protocol 28 and below. If it is 28 or below, it means the server is running 0.4.15, and unicodeparser will display the message to the player before sending it to the server.

0.4.16 and 0.4.17 uses protocol 32, and the upcoming 5.0.0 might be using protocol 36. The message will not be displayed by unicodeparser for those servers because they will display it by themselves. That is why I did protocol check, if I don't do the check then the message will be duplicated (one shown by unicodeparser and another shown by server).

Try to update unicodeparser now and see if it's working correctly.
